File: binary_tree/Java/MirrorImageBT.java
import java.util.*;

class Node {
    int data;
    Node  left,  right;
    Node(int data)
    {
        this.data=data;
        left=null;
        right=null;
    }
}
public class MirrorImageBT {
    static boolean isMirror(Node node1, Node node2){
        if(node1 == null && node2 == null){
            return true;
        }

        if(node1 == null || node2 == null){
            return false;
        }

        return (node1.data == node2.data) && isMirror(node1.left, node2.right) && isMirror(node1.right, node2.left);
    }

    public static void main(String[] args) {
        Node a = new Node(1);
        Node b = new Node(1);
        a.left = new Node(2);
        a.right = new Node(3);
        a.left.left = new Node(4);
        a.left.right = new Node(5);

        b.left = new Node(3);
        b.right = new Node(2);
        b.right.left = new Node(5);
        b.right.right = new Node(8);
        boolean res = isMirror(a, b);
        if(res){
            System.out.println("Both are Symmetric");
        }
        else{
            System.out.println("Not Symmetric");
        }
    }
}

File: binary_tree/python/MirrorImageBT.py
# Time: O(n)
# Space: O(h)

class Node:
    def __init__(self, val):
        self.val = val
        self.left = self.right = None

def is_mirror(node1, node2):
    if(node1 is None and node2 is None):
        return True

    if(node1 is None or node2 is None):
        return False

    return (node1.val == node2.val) and is_mirror(node1.left, node2.right) and is_mirror(node1.right, node2.left)

if __name__=='__main__':
    node1 = Node(1)
    node2 = Node(1)
    node1.left = Node(2)
    node1.right = Node(3)
    node1.left.left = Node(4)
    node1.left.right = Node(5)

    node2.left = Node(3)
    node2.right = Node(2)
    node2.right.left = Node(5)
    node2.right.right = Node(8)

    print(is_mirror(node1, node2))